lib/cache/OperationWriter.js

```diff
@@ -0,0 +1,255 @@
+    this.timeout = timeout;
+    // whether or not the operation writer is running
+    this.running = false;
+    // an ID for when the next timeout-based flush will occur
+    this.timeoutId = null;
+    // the queue of operations
+    this.queue = [];
+    // current stopped promise; resolves once stopped
+    this.stopped = null;
+    this._resolveStopped = null;
+    // tracks operation hashes for duplicate prevention
+    this.recentOperationHashes = new LRU({max: 1000, maxAge: 5 * 60 * 1000});
+    // operation queue collection
+    this.collection = database.collections.continuity2017_operation_queue;
+    // queue length in database
+    this.databaseOperationQueueLength = 0;
+    const operationsConfig = config['ledger-consensus-continuity'].operations;
+    // max operation queue length
+    this.MAX_OPERATION_QUEUE_LENGTH = operationsConfig.maxQueueSize;
+  }
+
+  /**
+   * Starts this OperationWriter. Any operations that have been added to its
+   * queue will be bulk inserted when the queue fills or when a timeout occurs.
+   */
+  start() {
+    const {queue, running, highWaterMark, timeout} = this;
+    if(running) {
+      // already running
+      return;
+    }
+
+    // now running
+    this.running = true;
+
+    // if the queue is at or over the highWaterMark, schedule immediate flush
+    if(queue.length >= highWaterMark) {
+      this._scheduleFlush({timeout: 0});
+    } else if(queue.length > 0) {
+      // queue is not empty, schedule a timeout flush
+      this._scheduleFlush({timeout});
+    }
+
+    this.stopped = new Promise(res => {
+      this._resolveStopped = res;
+    });
+  }
+
+  /**
+   * Stops this OperationWriter. Events can still be added to the queue, but
+   * they will not be automatically bulk inserted.
+   */
+  stop() {
+    const {running, timeoutId} = this;
+    if(!running) {
+      // already not running
+      return;
+    }
+    // cancel any scheduled flush
+    if(timeoutId !== null) {
+      clearTimeout(timeoutId);
+      this.timeoutId = null;
+    }
+    this.running = false;
+    this.stopped = null;
+    this._resolveStopped();
+  }
+
+  /**
+   * Adds an operation to be written. If this OperationWriter is running and the
+   * operation causes the queue to reach the configured `highWaterMark`, then
+   * it will automatically flush operations to the underlying database via a
+   * bulk insert.
+   *
+   * @param {object} options - Options to use.
+   * @param {object} options.operation - The operation to write.
+   * @param {object} options.meta - The metadata to write.
+   */
+  async add({operation, meta, forceFlush = false} = {}) {
+    const {queue, running, highWaterMark, MAX_OPERATION_QUEUE_LENGTH} = this;
+
+    if(this.databaseOperationQueueLength >= MAX_OPERATION_QUEUE_LENGTH) {
+      // we intentionally do not await the promise, the length can be updated
+      // after the error is returned
+      this._cacheDatabaseQueueLength().catch(e => {
```

Comment:
While this is doing caching, I think that's an internal detail of the function. What it's actually doing is updating the queue length, right? So I think it should be named that. I had to read all the code here to figure out why it was caching anything and went looking to see if the cache was ever read and got confused. I know understand that multiple calls may be made to update the queue length concurrently -- and you're trying to effectively debounce that. So, I think if you name this `_updateDatabaseQueueLength` it will be clearer what the point of the function is.

```suggestion
      this._updateDatabaseQueueLength().catch(e => {
```

Reply:
It would also be good to add a note here as to why the length gets updated in this case. I had to reason through why this would be called here. I saw that the only other place that the queue length gets updated is the intuitive spot where the ops get flushed to the database. It looks like we have no feedback from when the queue gets drained, so this is a place where we test to see if anything has drained because we're over the limit.

I'll note that if the queue has drained we'll still reject the current op because we throw without waiting for the fresh queue length. We should mention that we know that and it's intentional here (unless it wasn't) -- for simplicity of design. We should also mention that since we don't get any feedback on when the queue is draining, we make the update call to get a fresh value for the database queue length whenever we're at/over the limit.